_posts/2019-04-16-preservation.md

```diff
@@ -0,0 +1,358 @@
+   corrupted by a race condition, and versions will be unable to be properly
+   calculated:
+   1. Preservation starts. Pages start to persist into a bag.
+   2. Page gets modified out of band (label changed or page switched)
+   3. Page's changes get sent to be preserved.
+   4. Page gets preserved as a new version of the bag, updating the
+      `inventory.json` and incrementing the version counter.
+   5. Initial preservation ends. Updates inventory, but has out-of-date checksum
+      information.
+   6. However, since versions are numbered we could theoretically check to see
+      before we start writing things if files exist to reduce the time we need
+      to lock.
+
+#### Google Cloud Storage
+
+A small note about GCS - I'm going to call paths "directories", but in GCS it's
+  all just naming semantics. `foo/bar/txt.yml` isn't a file in the directory
+  `bar`, it's a file that has a name which happens to have slashes in it. The
+  GCS tools just visualize them as directories to be nice.
+
+1. Create a base directory named `<book-id>`
+1. Upload JSON for base object to `<book-id>/<book-id>.json`
+1. If there are any binary files attached, upload them to a `data` directory,
+   like `<book-id>/data/<file-id>-<file-name>.<file-extension>`
+1. For every child, recursively start at step 1, but start in a directory named
+   after its hierarchy. IE, for a page, it would go into
+   `<book-id>/data/<page-id>'
+
+**Pros**:
+
+1. Because there is no parent "manifest" that all children have to update
+   through, we can parallelize uploads of content to preservation no matter
+   where they go over an arbitrary number of machines.
+1. Race conditions can be handled via Figgy's optimistic locking mechanism for
+   each individual resource.
+1. Versioning and checksums are a property of GCS. We just enable it on the
+   bucket and don't have to do anything special.
+
+**Cons**:
+
+1. Prioritizing parallelization makes de-duplicating infeasible.
+2. The format's not standard. It's easy to look at, but we can't point at a
+   specification. At least it should be easy to convert to any of the other
+   formats if we'd like to.
+3. Migrating to a different platform is potentially more than a single copy
+   operation - but at least GCS validates file integrity as part of the
+   platform.
+
+### Delete a Full Book
+
+#### BagIt
+
+Either don't delete anything or store bags in GCS and delete the parent
+directory, relying on versioning.
+
+#### OCFL
+
+See Deletion in the [OCFL Implementation
+Notes](https://ocfl.io/0.2/implementation-notes/#forward-delta)
+
+1. Create a new `<book-id>/v2 directory.
+2. Create a `v2/inventory.json` file with with a new `version` entry that has
+   nothing in the `state` property.
+3. Create a new `v2/inventory.json.sha512`
+4. Copy the new `inventory.json` and `inventory.json.sha512` to the top
+   directory.
+
+There are unlikely to be locking problems, because nothing happens to the object
+post-delete.
```

Comment:
This is one way to delete an object, if you're ok with also keeping it. If you need to completely delete it for legal reasons there is another way outlined in the implementation notes. 

I wonder if PUL's collections would need this though (Emory's collections are more modern and this is a use case we've already had to deal with even lacking a comprehensive access system).

Reply:
We've had to do it in the past, is my understanding, it just wasn't the use case here.

Reply:
Agree —  we have had to takedown and permanently delete things, but it's the rare exception, not the typical delete.